### PR TITLE
feat: add the library text core — unicode fold, filename parse, author keys

### PR DESCRIPTION
**Pull request by oreglio**
**Part 1/5 of the Library view** — design, measurements and on-device photos in #2878 (now the tracking PR for the series).

> **Where this sits:** filenames are the *fallback*, not the method. With the shelf's
> "Use book metadata" option on, titles and authors come from inside the book —
> EPUB cache first, size-capped OPF read otherwise — in the index builder (part 3,
> #2887). These utilities are what the shelf falls back to for a book that carries
> no usable metadata, and they are deliberately conservative about it.

Pure, HAL-free text utilities the library index is built on:

- **Unicode fold** for sorting and matching: accents stripped, case folded, punctuation separating words, optional leading-article removal ("The Hobbit" ~ "hobbit"). Typographic dashes/quotes fold like their ASCII forms.
- **Filename parsing** — only for the ` -- ` export convention (Calibre, Anna's Archive and friends: `Title -- Author -- Year -- Publisher -- ISBN -- …`). A stem with no ` -- ` is kept **whole as the title** and yields no author: a single ` - ` is not treated as a separator, because plenty of real titles contain one. Of the ` -- ` segments only segment 1 is ever read, and it passes a guard that rejects years, ISBNs, hex digests and organisation credits ("Internet Archive", "Open Library") so they are not printed as people.
- **Author keys**: surname-first, so "Sally Rooney" and "Rooney, Sally" agree; used later for the one-spelling-per-author pass and the author sort.
- **Query matching** (word-prefix, folded) for the search feature.

**Failure mode is silence, not a wrong guess.** Anything the parser can't
confidently place stays in the title; a filename-derived author is later recorded
with provenance `UNKNOWN` in the index, so the UI never presents a guess as fact.

**How to review:** the host tests are the executable spec — `test/library_text` (19 tests). No firmware behavior changes; nothing consumes this yet.

**Series:** #2878 (tracking) → **this** → CLX1 format+reader → index builder → shelf activity → wiring.

### AI Usage
YES — see the disclosure in #2878; same process for every part of this series.

### CODE REVIEW

File: lib/LibraryIndex/LibraryText.cpp [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+  std::string out;
+  out.reserve(text.size());
+
+  const auto* cursor = reinterpret_cast<const unsigned char*>(text.data());
+  const auto* end = cursor + text.size();
+  bool pendingSpace = false;
+
+  while (cursor < end) {
+    const uint32_t cp = utf8NextCodepoint(&cursor);
+    if (cp == 0) break;
```
[the remark this answers is not in the document]
Comment (oreglio) [reply lib/LibraryIndex/LibraryText.cpp:175]: Good instinct, though not reachable as shipped: every call site in this series passes a `std::string` (whose `data()` is NUL-terminated since C++11), and the shared decoder stops on NUL — so the truncated-tail read cannot occur today. But `fold`'s contract is `string_view`, which is allowed to end mid-sequence, so I hardened it anyway: the loop now refuses to decode a lead byte whose promised continuation bytes lie past the view's end, with a regression test (`TruncatedTrailingSequenceIsDropped`). The shared `Utf8` decoder is upstream code this series doesn't touch.

File: lib/LibraryIndex/LibraryText.cpp [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+}
+
+bool allHexDigits(std::string_view s) {
+  if (s.empty()) return false;
+  for (const char c : s) {
+    const bool hex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f');
+    if (!hex) return false;
+  }
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (oreglio) [reply lib/LibraryIndex/LibraryText.cpp]: Fixed — a digest now requires at least one decimal digit (`looksLikeHexDigest`), so "Bede" and "Abba" survive as authors; tests added. The trade-off runs the other way for pure-letter strings like "deadbeef", which would now read as an (odd) author instead of being dropped — statistically a real hash fragment without a single digit is vanishingly rare, so this is the right side of the trade.

File: lib/LibraryIndex/LibraryText.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// multi-author separator, and the trailing underscores and punctuation that
+// exporters leave behind ("Qiu Xiaolong_", "Michael S_ Heiser"). An underscore
+// between letters becomes a full stop, since that is what it replaced in a name
+// a filesystem refused to hold.
+//
+// It deliberately does NOT swap "Last, First" into "First Last": "Qiu Xiaolong"
+// and "Lee Min Jin" defeat every such rule, and guessing wrong is worse than
+// leaving the author's own spelling alone. Harmonising the several spellings of
+// one person is done by picking the most common one that actually occurs, which
+// needs the whole library and so belongs to the index build.
```
[the remark this answers is not in the document]
Comment (oreglio) [reply lib/LibraryIndex/LibraryText.h]: All three fixed — the header now states what the code does: `cleanPersonName` inverts a single-comma name and leaves multi-comma names and lists alone; `authorKey` truncation is documented as byte truncation with the "alex michael" rationale; and the "no First/Last heuristic anywhere" claim is scoped to the author key, with the display-side rules named explicitly.

File: test/library_text/LibraryTextTest.cpp [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+TEST(LibraryAuthorKey, DistinctPeopleDoNotCollide) {
+  EXPECT_NE(authorKey("Alex Michaelides"), authorKey("Ashley Elston"));
+  EXPECT_NE(authorKey("Ian Manook"), authorKey("Min Jin Lee"));
+}
+
+TEST(LibraryAuthorKey, FitsTheRecordFieldWithoutCollapsingToAForename) {
+  const std::string key = authorKey("Bartholomew Fitzgerald Wellington");
+  EXPECT_LE(key.size(), library::AUTHOR_KEY_MAX_BYTES);
+  EXPECT_NE(key.back(), ' ');
```
[the remark this answers is not in the document]
Comment (oreglio) [reply test/library_text/LibraryTextTest.cpp:185]: Both fixed — the self-equality assertion is gone (the reversed-name inequality stays, which is what the test actually proves), and the authorKey test now asserts the key is non-empty before touching `.back()`.

**Comment by Uri-Tauber** [comment]
Before actually reviewing the code, I want to discuss the method the PR description explain. 
I think parsing the bare filename is very fragile. My library have a lot of books which (I think) will break this system.

Just think on this very common structure:
```
Hacker's Delight (2nd Edition) -- Henry S_ Warren, Henry S_ Warren Jr -- 2, 2012 -- Addison-Wesley Professional -- 9780133084993 -- eca63642edb16597640eab4abde1767d -- Anna’s Archive (1)
```

And some books include series name, some include the book name in two different languages... Basically, any parser will fail in a lot of cases.

So my question is why not pull the book/author/series from the epub metadata, like recent activity does today? 
I assume you already thought and rejected this for some reason (maybe the decompression of every epub took too long?).
It probably should be one bigger delay when the user open the library activity for the first time, and then get stored somewhere for easier fetching next time.

**Comment by oreglio** [comment]
The design does what you describe — the filename parse is the fallback layer, not the method. With "Use book metadata" enabled, the index builder (part 3, #2887) pulls title/author from inside the book: EPUB cache first when the book was opened before (the same metadata-only epub.load(false, true) Recent activity uses), otherwise a size-capped read of the OPF package document — two small zip entries, not a full decompression. Results are stored in the CLX1 index, so it's exactly the "one bigger delay, then stored" shape you suggest: a popup-fronted rebuild (~8 s for a 58-book card on an X3, including the OPF pass for never-opened books), then cheap reads afterwards.

On your example: this parser was actually built against that Anna's Archive shape. It splits on --, takes segment 0 as the title, and classifies only segment 1 through a guard rejecting years, ISBNs, hex digests and org credits — there's a test on precisely that structure (Sample Title -- Xun, Lu -- 2012 -- Sample Press -- d8fc -- Open Library). Your file parses as title "Hacker's Delight (2nd Edition)" with the Warren segment as author. When a filename doesn't match any known shape, degradation is the whole stem as title and no author — never a wrong-field guess — and filename-derived authors are recorded with provenance "unknown" in the index, so the UI never presents a guess as fact.

Metadata mode is opt-in rather than always-on because the OPF pass costs real time on SPI SD and some users prefer their curated filenames; happy to discuss flipping the default.

Series extraction is out of scope for this series.

I'll add a line to this PR's description making explicit that this is the fallback layer under the metadata reader in #2887.

**Comment by oreglio** [comment]
Correcting myself: I ran your exact filename through the parser rather than reasoning about it, and one of the claims in my previous comment doesn't survive that.

```
Hacker's Delight (2nd Edition) -- Henry S_ Warren, Henry S_ Warren Jr -- 2, 2012 -- Addison-Wesley Professional -- 9780133084993 -- eca63642edb16597640eab4abde1767d -- Anna's Archive (1)

  title        : "Hacker's Delight (2nd Edition)"     ok
  author (raw) : "Henry S_ Warren, Henry S_ Warren Jr" ok
  author shown : "Henry S Warren Jr Henry S Warren"    wrong
```

The title is right and the segment is picked correctly, but the displayed name is scrambled — **you found a real bug.** `cleanPersonName` turns "Austen, Jane" into "Jane Austen", because publishers ship both spellings and an author with a single book gives the spelling vote no majority to work with. Your file carries the duplicated-author artifact those exports leave behind, which contains exactly one comma, so the inversion fires on it.

I'd rather not teach the heuristic to recognise duplicates; that road has no end. The narrow fix is to let the inversion run only when the part after the comma is actually shaped like a given name — at most three words. "Jane" and "Herbert G." keep inverting, "Henry S Warren Jr" stops, and the name is then displayed exactly as the filename spells it instead of rearranged. The alternative is dropping the inversion altogether, which costs the "Austen, Jane" case and buys simplicity. I'll take whichever you prefer.

One clarification I owe you, since my PR description caused this whole detour. It claimed the parser handles ``Title - Author.epub``. It does not, and I've now corrected the description. The only separator is a **double** dash surrounded by spaces:

```
" -- "   separator, the Calibre / Anna's Archive export convention
" - "    not a separator — far too many real titles contain one
```

A stem with no `` -- `` is kept whole as the title and produces no author at all, which is what the `NoSeparatorKeepsTheWholeStemAsTitle` test locks in. So the fragility you were picturing — a single dash splitting arbitrary filenames — isn't there; I described the code badly and you reviewed the description, which is on me.

**Comment by muhas** [comment]
I'll just leave here how my files are organized, in case it comes in handy.

Screenshot_20260807-224428
my default structure folder/autor/serial/#number_title[blablabla]_bla.epub or folder/autor/title[blablabla]_blablabla.epub
and opds download: autor - title or title - autor

**Comment by oreglio** [comment]
Thanks — concrete structures are exactly what's useful here, and yours surfaced something I should raise myself rather than let you discover.

**What already works with your layout.** The flat shape (`folder/author/title[...].epub`) is handled: when nothing better is available, the book's parent folder becomes the author, so your `author` directory is read as the author rather than ignored. Bracketed and parenthesised spans are dropped from author names, and underscores between letters are read as the full stops a filesystem won't take (`Herbert G_ Wells` → `Herbert G. Wells`).

**Where your layout breaks it.** Your series shape, `folder/author/serial/#number_title.epub`, would show the *series* as the author, because the rule takes the immediate parent folder and that's the series directory. That's a real gap, and it's the same theme @Uri-Tauber raised about series in filenames. Walking up a level when the immediate parent looks like a series is the obvious direction; it isn't implemented.

**On `author - title` from OPDS.** Filename parsing only splits on ` -- ` (a double dash surrounded by spaces), never on a single dash. That's deliberate rather than an omission: single dashes live inside titles and names constantly — *Spider-Man*, *Jean-Paul*, *Anne-Marie* — so splitting on them would corrupt far more names than it would recover. With the double dash the separator is unambiguous.

**The part that matters most, and it's a limitation not a feature.** If your library is not in the Latin alphabet, most of this is moot in a way I should state plainly: the text normalisation used for sorting and search drops any script it doesn't map. I checked rather than assumed — `Война и мир` normalises to an empty string, as do Greek and CJK titles, while `L'Énéide` correctly becomes `eneide`. The practical effect is that non-Latin titles are not searchable, sort together under an empty key, and can't be reached from the A-Z grid. Enabling "Use book metadata" gets you the correct title and author *displayed* from inside the EPUB, but it does not fix search or sorting, because those run on the normalised form.

That's fixable — keeping unmapped scripts in the normalised string instead of discarding them would restore search and sorting within a script — but it isn't done today, and I'd rather say so up front. If your library is largely non-Latin, I'd genuinely like to know: it changes how much this feature is worth to you as it stands.

**Comment by muhas** [comment]
> If your library is largely non-Latin,

Yes, basically reading native or translated literature and the file names are the same... Their search at least by file/directory name would already be good (If, of course, switching keyboard layouts appears). In a few days I will be able to check how it works in general on a large library with a similar structure (as example: https://disk.yandex.ru/d/y3psCh6IdbOnuw), I don’t know if indexing will be able to cope with such a volume of books

**Comment by Uri-Tauber** [comment]
Hmm. I dont know...
I would prefer to drop out any filename parsing. There are simply too many edge cases, which can be a maintenance burden later. Reading only OPF metadata is the path Calibre and Koreader took, so I dont think we should over-engineer this.

However, Its a design choice and I'm open to other opinions here.

**Comment by oreglio** [comment]
Fair point, and the default is worse than you're describing: metadata sits behind a toggle that's off, so parsing is what everyone actually gets. I'll flip that. OPF by default, filename only for what the OPF leaves empty.

What I'd keep is that fallback. @muhas's library upthread is the case: non-Latin filenames carrying the only usable information, OPF often absent or wrong. Without a fallback those books land on the shelf unnamed — worse than the folder browser this is meant to improve on.

On the maintenance worry: it's one file (`LibraryText`), no HAL dependency, 196 host-test assertions in CI. An edge case costs a test, not a device session.

The whole surface is four rules: ` -- ` splits `author -- title` (never a single dash); brackets dropped from author names; `_` between letters reads as `.`; parent folder becomes the author below top level. Tell me which you don't want
to own and I'll delete them in this PR — starting with the fourth, which is what shows the *series* as the author in `author/series/#01_title.epub`.

**Comment by Uri-Tauber** [comment]
@oreglio I'm in favour of deleting the entire mechanism. No other ereader I know try to parse the file names. 
Similar example, Music player apps also read the metadata and dont try to parse the filename. 

The filename should be a fallback for bookname, if the metadata doesnt contain it. Autor name should be extracted only from the metadata. If none available, group under **Unknown**.

**Comment by oreglio** [comment]
Works for me, deleting it. Author from metadata only, filename as the title fallback when the metadata has none, everything without an author under **Unknown**. The music-player comparison is the right one.

Pushed to #2885, the branch is a single commit again, −274 lines, 32 host tests green. The Unknown grouping itself lands with the index builder and the shelf, in #2887 and #2888.

**Comment by Uri-Tauber** [comment]
Sorry for the wait. Been very busy with other PRs. 
Overall the code look clean, and I'm fine with it. Two small optimizations can be made.

One thing I'm concerned about is we want to do new release soon, and we wont have the time to review your entire series. So, to keep half-baked feature out of the release, I'm thinking to create special branch and direct your PRs to it, and when the feature will be complete I will merge it to `develop`. Are you OK with that?

BTW, Justin made huge refactor to how lists are created, so you will probably have a lot of conflict-solving to do.

File: lib/LibraryIndex/LibraryText.cpp [hunk 5]
```diff
@@ -0,0 +1,351 @@
+#include "LibraryText.h"
+
+#include <Utf8.h>
+#include <Utf8ComposeTable.h>
```
Comment (Uri-Tauber) [inline lib/LibraryIndex/LibraryText.cpp]: including the static compose table creates a second ~5 KB flash copy. 

Consider expose reverse lookup from Utf8 so both users share one table.

File: lib/LibraryIndex/LibraryText.cpp [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+      continue;
+    }
+    if (c == ';') break;
+    if (depth == 0) cleaned.push_back(c);
+  }
+
+  const std::string folded = fold(cleaned);
+
+  constexpr size_t MAX_TOKENS = 12;
+  std::string tokens[MAX_TOKENS];
```
Comment (Uri-Tauber) [inline lib/LibraryIndex/LibraryText.cpp]: Twelve live `std::strings` allocate while building every author key and add ~288 B stack pressure. Sort bounded `string_views`/offsets into folded text instead.

**Comment by aurelien-edusign** [comment]
No problem.
Both applied. The compose table lives in one TU now, reverse lookup moved to`Utf8` as `utf8DecomposedBase()`  and that file alone went 16062 -> 9300 bytes on a host -Os build. `authorKey()` sorts `string_view`s; a sibling function had
twenty-four `std::string`s, 576 B of stack, same fix.

Feature branch: yes, point them wherever suits you. Do you keep it in sync with `develop`, or should I rebase as it moves?

On Justin's refactor — the shelf can take `UiListActivity` for input, `ListNav` and chrome. It still needs its own row builder: the label band is one line high whenever a subtitle is present, and the shelf wraps a title over three lines
with the author beneath.

One for you: exporters often write `dc:creator` as the literal "Unknown". Those group under **Unknown** too, or show as an author named Unknown? Merging them costs one normalisation on metadata.

**Comment by Uri-Tauber** [comment]
@oreglio I think you replied from the wrong github account. :smile:

Regarding the feature branch, I'm not sure. Will see how many conflicts pop up...

About the refactor, yes. Use freeInk components, they automatically support touch screens, which is very important.
If your next PRs add new component, consider moving to to freeInk.

About "Unknown" author, if all it take is just one comparison, I support merging it.

**Comment by oreglio** [comment]
oops I forgot to switch from my work account 🤦‍♂️

Thanks for the merge. I'll retarget #2887-#2889 to `feat/library-view` #2886 you already moved.

Unknown: it is one comparison, so merged. A `dc:creator` of literally "Unknown" now lands in the same bucket as a missing one.

On FUI, agreed, and touch is the argument that settles it. The one thing the shelf needs that `fui::list` can't express today is a row whose label wraps over three lines with the author beneath it: the label band is one line high whenever a subtitle is present. Per your last point I'd rather add that to freeInk than hand-draw it here. Do you want the component, or should the shelf ship with a single-line title first and gain it later?

**Comment by Uri-Tauber** [comment]
@oreglio I didnt understood your description of the component.
Can you share a screenshot or something?

**Comment by oreglio** [comment]
I meant like this one 

CleanShot 2026-08-14 at 00 20 05@2x


One row per book: the title wraps over as many as three lines, and the author sits underneath it. "The Mom Test: how to talk to customers and learn if you business is a good idea when everybody is lying to you" is one book, not three.
Real titles from a real card, that's where three lines comes from.

`fui::list` gives a row's title a single line as soon as that row also carries a subtitle. Through it, this screen would read "The Mom Test: how to talk to cust…" and lose the rest. Sorted by author it's fine, because the author becomes
a heading and the row has no subtitle, so it's this view that needs it.

What I have in mind is one property on a row: how many lines its title may use, with the row growing to fit them and the subtitle. If that sounds right I'll open a PR on freeInk for it, and the shelf uses it.

The render for by Author is this one 
CleanShot 2026-08-14 at 00 26 31@2x

**Comment by Uri-Tauber** [comment]
Hmmm.

I think having the names truncated via `fui::list` is acceptable for now. Personally, I would prefer to have more books in the screen, even if half the name is hidden.

It can always be changed in the future.

**Comment by oreglio** [comment]
Agreed, and done that way: I've pushed the converted shelf to #2888. Rows through `fui::list`, the sort strip through `fui::tabBar`, the A-Z grid as fui buttons, so touch comes for free, with titles truncated to one line, exactly your call. More books per screen it is. CI is green on the branch; I'm keeping the PR in draft until #2886/#2887 land so its diff shows only the shelf (~1.3k lines, two commits: the activity, then its conversion).

On "it can always be changed in the future": that future now has a vehicle upstream, Free-Ink/freeink-sdk#37 teaches `fui::list measured` per-item heights for wrapped labels above a subtitle, composing with Justin's subtitle-wrapping commit. If it lands, re-enabling wrapped titles here becomes a one-line maxLines change, whenever (and if) you want it.
